torch/_functorch/vmap.py

```diff
@@ -152,20 +152,9 @@ def incompatible_error():
             f'out_dims has structure {tree_flatten(out_dims)[1]} but outputs '
             f'has structure {output_spec}.')
 
-    if isinstance(batched_outputs, torch.Tensor):
-        # Some weird edge case requires us to spell out the following
-        # see test_out_dims_edge_case
-        if isinstance(out_dims, int):
-            flat_out_dims = [out_dims]
-        elif isinstance(out_dims, tuple) and len(out_dims) == 1:
-            flat_out_dims = out_dims
-            out_dims = out_dims[0]
-        else:
-            incompatible_error()
-    else:
-        flat_out_dims = _broadcast_to_and_flatten(out_dims, output_spec)
-        if flat_out_dims is None:
-            incompatible_error()
+    flat_out_dims = _broadcast_to_and_flatten(out_dims, output_spec)
+    if flat_out_dims is None:
+        incompatible_error()
```

Comment:
> `_broadcast_to_and_flatten([1], *)` should return `[1]` but instead returns `None`

What is the argument for why this should be the behavior?

I think `test_out_dims_edge_case` is wrong and we should not be allowing that behavior, but I do not remember why we added the special case for it.

Morally, in `_broadcast_to_and_flatten(a, b)`, I'd expect b's tree-structure to be "larger" than a's tree-structure, and this is not the case with `_broadcast_to_and_flatten([1], *)`

Reply:
My goal was to remove the edge case from `torch/_functorch/vmap.py:155`

Reply:
I see, and the edge case is making it difficult to add support for `None` returns?

Directionally we should probably remove the edge case - it looks like JAX does not support it, so we shouldn't either.

In order to remove it, I'd be OK with deprecating it in a release and deleting it in the subsequent release. Open question to if we can just delete the edge case in a single release and label it as BC-breaking. If we want to go down this path, then we should do some code archeology:
1) why was the edge case added in the first place? I don't remember. It's possible that either (a) JAX used to support it and we copied the behavior, or (b) I got confused about what the output of a Python function is.
2) Try to browse usages of functorch.vmap on GitHub (maybe via cs.github.com?). There should be at most ~300 of them right now. See if any do rely on this edge case and if we're breaking anyone.

Off of the top of my head I don't think anyone really uses the out_dims flag, so it might be safe to delete the edge case in a single release.

Reply:
> I see, and the edge case is making it difficult to add support for None returns?

It is not 'difficult' I just needed to add more logic to the edge case handling, so I thought I would try to remove it first.